File: js/merge_sort.js
const arr = [3, 2, 44, 2315, 12, 4, 8, 7, 0, -12];
const mergeSort = (arr) => {
        if (arr.length < 2) return arr;
        const mid = Math.floor(arr.length / 2);
        return merge(mergeSort(arr.slice(0, mid)), mergeSort(arr.slice(mid)));
}

const merge = (a, b)=>{
        let result = [];
        let i = 0;
        let j = 0;

        while(i < a.length && j < b.length) {
                if (a[i] < b[j]) {
                        result.push(a[i]);
                        i++;
                }
                else {
                        result.push(b[j]);
                        j++;
                }
        }
        while(i < a.length){
                result.push(a[i]);
                i++;
        }
        while(j < b.length){
                result.push(b[j]);
                j++;
        }
        return result;
}

console.log(arr);
console.log(mergeSort(arr));

File: snakes/merge_sort.py
import math

sortThis = [2,3,2,1,0]
print('this is supposed to merge sort this: \n')
print(sortThis)

def mergeSort(array):
	if len(array) > 1:
		if len(array) > 1:
			mid = len(array)//2
			left = array[:mid]
			right = array[mid:]

			mergeSort(left)
			mergeSort(right)

			a = 0
			b = 0
			c = 0
			lenLeft = len(left)
			lenRight = len(right)

			while a < lenLeft and b < lenRight:
				if left[a] < right[b]:
					array[c] = left[a]
					a = a + 1
				else:
					array[c]=right[b]
					b = b + 1
				c = c + 1
			while a < lenLeft :
				array[c] = left[a]
				a = a + 1
				c = c + 1
			while b < lenRight:
				array[c] = right[b]
				b = b + 1
				c = c + 1

mergeSort(sortThis)
print("I sorted this: ", sortThis) 
